File: bionic-js-runtime-swift/Bjs/BjsContext.swift
import JavaScriptCore

class BjsContext {
    
    public let jsContext: JSContext = JSContext()
    public let projectName: String
    public var moduleLoader: JSValue!
    public var timeoutIds: NSMutableSet = NSMutableSet()
    public var intervalIds: NSMutableSet = NSMutableSet()
    public var lastTimeoutId: Int = 0
    public var lastIntervalId: Int = 0
    var nativeWrappers = [String : BjsNativeWrapper.Type]()
    
    init(_ projectName: String) {
        self.projectName = projectName
        
        jsContext.exceptionHandler = { [weak self] context, exception in
            let bjsException = BjsException(exception)
            if bjsException.notLocalizable {
                return;
            }
            self!.logError(bjsException.description)
            context?.exception = exception
        }
        
        let consoleLog: @convention(block) (_ : String) -> Void = { [weak self] in
            self!.logInfo($0)
        }
        let consoleError: @convention(block) (_ : String) -> Void = { [weak self] in
            self!.logError($0)
        }
        let console = jsContext.objectForKeyedSubscript("console")!
        console.setObject(consoleLog, forKeyedSubscript: "log" as NSString)
        console.setObject(consoleError, forKeyedSubscript: "error" as NSString)
        
        let setTimeout: @convention(block) (_ : JSValue, _ : Int) -> Int = { [weak self] in
            let handlerId = self!.newTimeoutId()
            self!.runTimeout(handlerId, $0, $1)
            return handlerId
        }
        jsContext.setObject(setTimeout, forKeyedSubscript: "setTimeout" as NSString)
        
        let clearTimeout: @convention(block) (_ : Int) -> Void = { [weak self] in
            self!.clearTimeout($0)
            
        }
        jsContext.setObject(clearTimeout, forKeyedSubscript: "clearTimeout" as NSString)
        
        let setInterval: @convention(block) (_ : JSValue, _ : Int) -> Int = { [weak self] in
            let handlerId = self!.newIntervalId()
            self!.runInterval(handlerId, $0, $1)
            return handlerId
        }
        jsContext.setObject(setInterval, forKeyedSubscript: "setInterval" as NSString)
        
        let clearInterval: @convention(block) (_ : Int) -> Void = { [weak self] in
            self!.clearInterval($0)
        }
        jsContext.setObject(clearInterval, forKeyedSubscript: "clearInterval" as NSString)
        
        // Shim for "process" global variable used by Node.js
        let process = JSValue.init(newObjectIn: jsContext)
        process!.setObject(JSValue.init(newObjectIn: jsContext), forKeyedSubscript: "env" as NSString)
        jsContext.setObject(process, forKeyedSubscript: "process" as NSString)
        
        let bjsNativeRequire: @convention(block) (_ : String) -> JSValue = { [weak self] in
            return self!.getNativeModule($0)
        }
        jsContext.setObject(bjsNativeRequire, forKeyedSubscript: "bjsNativeRequire" as NSString)
        
        let bjsSetModuleLoader: @convention(block) (_ : JSValue) -> Void = { [weak self] in
            self!.moduleLoader = $0
        }
        jsContext.setObject(bjsSetModuleLoader, forKeyedSubscript: "bjsSetModuleLoader" as NSString)
    }
    
    func addNativeWrappers(_ nativeWrapperClass: BjsNativeWrapper.Type) {
        
        if nativeWrapperClass.bjsLocator.isInvalid {
            fatalError("invalid module locator")
        }
        
        let wrapperName = nativeWrapperClass.bjsLocator.moduleName
        if nativeWrappers[wrapperName] != nil {
            fatalError("native wrapper \"\(wrapperName)\" was already added to this Bjs context")
        }
        nativeWrappers[wrapperName] = nativeWrapperClass
    }
    
    func getModule(_ moduleName: String) -> JSValue! {
        return moduleLoader.call(withArguments: [moduleName])
    }
        
    private func getNativeModule(_ nativeModuleName: String) -> JSValue {
        return self.nativeWrappers[nativeModuleName]!.bjsGetNativeFunctions(self)!
    }
    
    private func newTimeoutId() -> Int {
        self.lastTimeoutId += 1
        let handlerId = self.lastTimeoutId
        self.timeoutIds.add(handlerId)
        return handlerId
    }
    
    private func newIntervalId() -> Int {
        self.lastIntervalId += 1
        let handlerId = self.lastIntervalId
        self.intervalIds.add(handlerId)
        return handlerId
    }
    
    private func runTimeout(_ handlerId: Int, _ function: JSValue, _ delayInMs : Int) {
        DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(delayInMs)) {
            if self.timeoutIds.contains(handlerId) {
                function.call(withArguments: [])
                self.timeoutIds.remove(handlerId)
            }
        }
    }
    
    private func runInterval(_ handlerId: Int, _ function: JSValue, _ delayInMs : Int) {
        Timer.scheduledTimer(withTimeInterval: Double(delayInMs) / 1000.0, repeats: true) { timer in
            if self.intervalIds.contains(handlerId) {
                function.call(withArguments: [])
            } else {
                timer.invalidate()
            }
        }.fire()
    }
    
    private func clearTimeout(_ handlerId: Int) {
        if self.timeoutIds.contains(handlerId) {
            self.timeoutIds.remove(handlerId)
        }
    }
    
    private func clearInterval(_ handlerId: Int) {
        if self.intervalIds.contains(handlerId) {
            self.intervalIds.remove(handlerId)
        }
    }
    
    func createJsObject(_ sourceObj: Any!) -> JSValue! {
        let jsObj = JSValue(object: sourceObj, in: jsContext)
        return jsObj
    }
    
    func createJsNull() -> JSValue! {
        return JSValue(nullIn: jsContext)
    }
    
    func createNativeExports() -> BjsNativeExports {
        return BjsNativeExports(jsContext)
    }
    
    func setJsExceptionToThrow(_ exceptionMessage: String) {
        jsContext.exception = JSValue.init(newErrorFromMessage: exceptionMessage, in: jsContext)
    }
    
    func getJsException() -> JSValue? {
        return jsContext.exception
    }
    
    func executeJs(_ code: String) -> JSValue! {
        return jsContext.evaluateScript(code)
    }
    
    func executeJs(_ code: String, _ filePath: String) -> JSValue! {
        return jsContext.evaluateScript(code, withSourceURL:URL(fileURLWithPath: filePath))
    }
    
    func logError(_ message: String) {
        print("Bjs \"\(projectName)\" error: \(message)")
    }
    
    func logInfo(_ message: String) {
        print("Bjs \"\(projectName)\" info: \(message)")
    }
    
    func getJsProp(_ obj: JSValue) -> [String] {
        let getOwnPropertyNames = obj.context.objectForKeyedSubscript("Object").objectForKeyedSubscript("getOwnPropertyNames")
        return getOwnPropertyNames?.call(withArguments: [obj]).toArray() as! [String]
    }
}

File: bionic-js-runtime-java/src/main/java/bionic/js/BjsContext.java
package bionic.js;

import edu.umd.cs.findbugs.annotations.NonNull;
import jjbridge.api.runtime.JSReference;
import jjbridge.api.runtime.JSRuntime;
import jjbridge.api.value.JSBoolean;
import jjbridge.api.value.JSDate;
import jjbridge.api.value.JSDouble;
import jjbridge.api.value.JSExternal;
import jjbridge.api.value.JSFunction;
import jjbridge.api.value.JSInteger;
import jjbridge.api.value.JSObject;
import jjbridge.api.value.JSString;
import jjbridge.api.value.JSType;
import jjbridge.api.value.JSValue;
import jjbridge.api.value.strategy.FunctionCallback;

import java.util.Date;
import java.util.HashMap;
import java.util.Map;

class BjsContext
{
    private final JSRuntime runtime;
    private final String projectName;
    private final TimeoutHandler timeoutHandler = new TimeoutHandler(0);
    private final Map<String, BjsNativeWrapperTypeInfo<?>> nativeWrappers = new HashMap<>();
    private JSReference moduleLoader;

    protected final FunctionCallback<JSReference> setTimeoutCallback = jsReferences ->
    {
        JSFunction<?> function = resolve(jsReferences[0]);
        int delay = ((JSInteger) resolve(jsReferences[1])).getValue();
        int timeoutId = timeoutHandler.runDelayed(function, jsReferences[0], delay);
        return newInteger(timeoutId);
    };

    protected final FunctionCallback<JSReference> clearTimeoutCallback = jsReferences ->
    {
        int id = ((JSInteger) resolve(jsReferences[0])).getValue();
        timeoutHandler.remove(id);
        return createJsUndefined();
    };

    protected final FunctionCallback<JSReference> bjsNativeRequireCallback = jsReferences ->
    {
        String moduleName = ((JSString) resolve(jsReferences[0])).getValue();
        return getNativeModule(moduleName);
    };

    protected final FunctionCallback<JSReference> bjsSetModuleLoaderCallback = jsReferences ->
    {
        this.moduleLoader = jsReferences[0];
        return createJsUndefined();
    };

    protected static void defineGlobalFunction(JSRuntime runtime, String name, FunctionCallback<JSReference> callback)
    {
        JSReference functionReference = runtime.newReference(JSType.Function);
        JSFunction<?> jsFunction = runtime.resolveReference(functionReference);
        jsFunction.setFunction(callback);
        runtime.globalObject().set(name, functionReference);
    }

    BjsContext(@NonNull JSRuntime runtime, @NonNull String projectName)
    {
        this.runtime = runtime;
        this.projectName = projectName;

        defineGlobalFunction(runtime, "setTimeout", setTimeoutCallback);
        defineGlobalFunction(runtime, "clearTimeout", clearTimeoutCallback);
        defineGlobalFunction(runtime, "bjsNativeRequire", bjsNativeRequireCallback);
        defineGlobalFunction(runtime, "bjsSetModuleLoader", bjsSetModuleLoaderCallback);

        //TODO: add console.log and console.error?

        // Shim for "process" global variable used by Node.js
        JSReference processRef = runtime.newReference(JSType.Object);
        JSObject<?> jsProcess = runtime.resolveReference(processRef);
        jsProcess.set("env", runtime.newReference(JSType.Object));
        runtime.globalObject().set("process", processRef);
    }

    <B extends BjsNativeWrapper<T>, T extends BjsExport> void addNativeWrapper(Class<B> nativeWrapperClass)
    {
        BjsNativeWrapperTypeInfo<B> typeInfo = BjsNativeWrapperTypeInfo.get(nativeWrapperClass);
        BjsLocator bjsLocator = typeInfo.bjsLocator;
        if (bjsLocator.isInvalid())
        {
            throw new RuntimeException("Invalid module locator");
        }

        String wrapperName = bjsLocator.moduleName;
        if (nativeWrappers.containsKey(wrapperName))
        {
            throw new RuntimeException("Native wrapper " + wrapperName + " was already added to this Bjs context");
        }
        nativeWrappers.put(wrapperName, typeInfo);
    }

    JSReference getModule(String moduleName)
    {
        return callFunction(moduleLoader, moduleLoader, newString(moduleName));
    }

    JSReference getNativeModule(String nativeModuleName)
    {
        return nativeWrappers.get(nativeModuleName).bjsGetNativeFunctions(this);
    }

    JSReference createJsNull()
    {
        return runtime.newReference(JSType.Null);
    }

    JSReference createJsUndefined()
    {
        return runtime.newReference(JSType.Undefined);
    }

    BjsNativeExports createNativeExports()
    {
        return new BjsNativeExports(runtime);
    }

    <T extends JSValue> T resolve(JSReference jsReference)
    {
        return runtime.resolveReference(jsReference);
    }

    JSReference newBoolean(boolean value)
    {
        JSReference reference = runtime.newReference(JSType.Boolean);
        ((JSBoolean) resolve(reference)).setValue(value);
        return reference;
    }

    JSReference newInteger(int value)
    {
        JSReference reference = runtime.newReference(JSType.Integer);
        ((JSInteger) resolve(reference)).setValue(value);
        return reference;
    }

    JSReference newDouble(double value)
    {
        JSReference reference = runtime.newReference(JSType.Double);
        ((JSDouble) resolve(reference)).setValue(value);
        return reference;
    }

    JSReference newString(String value)
    {
        JSReference reference = runtime.newReference(JSType.String);
        ((JSString) resolve(reference)).setValue(value);
        return reference;
    }

    JSReference newDate(Date value)
    {
        JSReference reference = runtime.newReference(JSType.Date);
        JSDate<?> jsDate = resolve(reference);
        jsDate.setValue(value);
        return reference;
    }

    <T> JSReference newExternal(T value)
    {
        JSReference reference = runtime.newReference(JSType.External);
        JSExternal<T> resolved = resolve(reference);
        resolved.setValue(value);
        return reference;
    }

    JSReference newFunction(FunctionCallback<?> callback)
    {
        JSReference reference = runtime.newReference(JSType.Function);
        JSFunction<?> jsFunction = resolve(reference);
        jsFunction.setFunction(callback);
        return reference;
    }

    JSReference newObject()
    {
        return runtime.newReference(JSType.Object);
    }

    JSReference newArray()
    {
        return runtime.newReference(JSType.Array);
    }

    JSReference callFunction(JSReference jsFunction, JSReference receiver, JSReference... arguments)
    {
        JSFunction<?> function = resolve(jsFunction);
        return function.invoke(receiver, arguments);
    }

    JSReference executeJs(String code)
    {
        return runtime.executeScript(code);
    }

    JSReference executeJs(String code, String filePath)
    {
        return runtime.executeScript(filePath, code);
    }

    void logError(String message)
    {
        System.err.println("Bjs \"" + projectName + "\" error: " + message);
    }

    void logInfo(String message)
    {
        System.out.println("Bjs \"" + projectName + "\" info: " + message);
    }
}
